src/librustc/ty/sty.rs

```diff
@@ -254,15 +257,38 @@ pub enum TypeVariants<'tcx> {
 /// handle). Plus it fixes an ICE. :P
 #[derive(Copy, Clone, PartialEq, Eq, Hash, Debug, RustcEncodable, RustcDecodable)]
 pub struct ClosureSubsts<'tcx> {
-    /// Lifetime and type parameters from the enclosing function.
+    /// Lifetime and type parameters from the enclosing function,
+    /// concatenated with the types of the upvars.
```

Comment:
But couldn't the types with those lifetimes be stored *somewhere else*, keyed by `DefId`?
I believe we already do something like that for the signature of the closure.